File: 3.LongestSubstringWithoutRepeatingCharacters/s.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        if(s.length() == 0) return 0;
        HashMap<Character, Integer> map = new HashMap<Character, Integer>();
        int max = 0;
        for(int i = 0, j = 0; i < s.length(); ++i){
            /*char already in the map*/
            if(map.containsKey(s.charAt(i))){
                /*move the left pointer to the pos where the repeated char found in the last time
                  i: the right pointer to scan the string
                  j: the left pointer to identify where the new substring starts*/
                j = Math.max(j, map.get(s.charAt(i))+1)  ; 
            }
            /*if char not exists in the map
              add char to the map, update max val*/
            map.put(s.charAt(i),i);
            max = Math.max(max, i-j+1);
        }
        return max;
    }
}

File: 3.LongestSubstringWithoutRepeatingCharacters/s.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        start = max_len = 0
        seen = {}
        for key, value in enumerate(s):
            # found a repeated value
            if value in seen:
                # update start, which is the index where sub-string starts
                # update only when the index of repeated value in seen is larger than start
                # which means if the index is smaller, we do not update start
                tmp = seen[value] + 1
                if tmp > start:
                    # update the start as one more larger than the new index
                    start = tmp

            max_len = max(max_len, key - start + 1)
                # add this char into seen, no matter it is repeated or not
            seen[value] = key
        # return when iteration is done        
        return max_len